app/backend/approaches/chatreadretrieveread.py

```diff
@@ -69,22 +69,37 @@ def __init__(
         self.chatgpt_token_limit = get_token_limit(chatgpt_model)
 
     async def run_until_final_call(
-        self, history: list[dict[str, str]], overrides: dict[str, Any], should_stream: bool = False
+            self,
+            history: list[dict[str, str]],
+            overrides: dict[str, Any],
+            should_stream: bool = False,
+            query_history: Optional[list[dict[str, str]]] = None
     ) -> tuple:
+
         has_text = overrides.get("retrieval_mode") in ["text", "hybrid", None]
         has_vector = overrides.get("retrieval_mode") in ["vectors", "hybrid", None]
         use_semantic_captions = True if overrides.get("semantic_captions") and has_text else False
         top = overrides.get("top") or 3
         exclude_category = overrides.get("exclude_category") or None
-        filter = "category ne '{}'".format(exclude_category.replace("'", "''")) if exclude_category else None
+        use_query_history = overrides.get("use_query_history", True)
+        query_filter = "category ne '{}'".format(exclude_category.replace("'", "''")) if exclude_category else None
+
+        query_history = query_history or []
+
+        if use_query_history:
+            query_history_input = query_history or []
+        else:
+            query_history_input = history
 
-        user_q = "Generate search query for: " + history[-1]["user"]
+        raw_user_query = history[-1]["user"]
+
+        user_q = "Generate search query for: " + raw_user_query
 
         # STEP 1: Generate an optimized keyword search query based on the chat history and the last question
         messages = self.get_messages_from_history(
             self.query_prompt_template,
             self.chatgpt_model,
-            history,
+            query_history_input,
```

Comment:
Our prompt says "Below is a history of the conversation so far, and a new question asked by the user that needs to be answered by searching in a knowledge base about employee healthcare plans and the employee handbook."
I'm surprised you got good results by passing in the query history since it would seem to be in disagreement with the prompt. You didn't need to alter the prompt at all?

Reply:
Hmm, yea our prompt is different, customizable for every use case. But is it in a disagreement with the this prompt? I don't think so, since the structure contains history but only in form of query messages and bot generated queries. This is basically list of few shots.